pyfr/plugins/sampler.py

```diff
@@ -5,51 +5,36 @@
 from pyfr.quadrules import get_quadrule
 
 
-def _closest_pts_bf(epts, pts):
-    for p in pts:
-        # Compute the distances between each point and p
-        dists = [np.linalg.norm(e - p, axis=2) for e in epts]
-
-        # Get the index of the closest point to p for each element type
-        amins = [np.unravel_index(np.argmin(d), d.shape) for d in dists]
-
-        # Dereference to get the actual distances
-        dmins = [d[a] for d, a in zip(dists, amins)]
-
-        # Find the minimum across all element types
-        yield min(zip(dmins, range(len(epts)), amins))
-
-
-def _closest_pts_kd(epts, pts):
-    from scipy.spatial import cKDTree
+def _closest_pts(epts, pts):
+    from rtree import index
 
-    # Flatten the physical location arrays
     fepts = [e.reshape(-1, e.shape[-1]) for e in epts]
 
-    # For each element type construct a KD-tree of the upt locations
-    trees = [cKDTree(f) for f in fepts]
+    idx = []
 
-    for p in pts:
-        # Query the distance/index of the closest upt to p
-        dmins, amins = zip(*[t.query(p) for t in trees])
+    for f in fepts:
+        pr = index.Property()
+        pr.dimension = f.shape[1]
+        pr.interleaved = True
+        insl = []
+        for i in range(f.shape[0]):
+            ins = tuple(np.concatenate([f[i,:], f[i,:]]))
+            insl.append((i,ins,None))
+        idx.append(index.Index(insl,properties=pr))
 
-        # Unravel the indices
+    for p in pts:
+        dmins = []
+        amins = []
+        for i, v in enumerate(idx):
+            amin = list(v.nearest(np.concatenate([p, p]), 1))[0]
```

Comment:
Why is list needed here?  If it is an iterator `next(...)` should get the first item in the sequence.  Also see if you can one-line amins and do dmins as a separate pass